### ISSUE TITLE
Duplicate symbols when building a busybox-like tool

### ISSUE THREAD

**Issue by vadimkantorov** [emscripten-core/emscripten#12167]
I'm building a xetex+dvipdfmx busybox-like two-in-one executable. I redefined names of main functions using compiler `-Dmain=...` argument.

I have:
```
libxetex.a
  someobjectfile.o jpegimage.o oldxetexmain.o (contains busymain_xetex)

libdvidpfmx.a
  someotherobjectfile.o jpegimage.o main.o olddvipdfmxmain.o (contains busymain_dvipdfmx)
```

Both `jpegimage.o` functions are used in `someobjectfile.o`/`someotherobjectfile.o` and export same-named symbols which causes duplicate symbols. Unfortunately the two copies of jpegimage.o are not guaranteed to be exactly same.

At the end of the day I'm interested to only link to `busymain_*` functions and hide all other symbols. Is there a way to postproc the static library `*.a` files and hide all unnecessary symbols prior to linking to the final `main.c`?

Thank you!

**Comment by sbc100** [comment]
When linking `.a` archive only the objects that are actually needed should be pulled in.  Does the same build work on your host platform (linux, mac, windows etc)?   This linking behaviour of emscripten is the same as on those platforms with respect the `.a` archive.

If both copies of `jpegimage.o` are being pulled in then there must be some symbol from the second copy that is not present in the first copy (otherwise the second copy would no be linked in).

(caveat: Unless you are building with `-s MAIN_MODULE`.. are you doing that?)

Having two different versions of `jpegimage` linked into the same program seems tricky at best and not a great idea.    The best approach is probably going to be trying to remove the duplicate versions or at least ensure they are the same version.   The second approach would be use `wasm-ld -r` (or `emcc -r`) to prelink each executable to a new, single object file and figure out a way to only export the main entry points from each one.  Then your final link would be `emcc main.o busymain_xetex.o busymain_dvipdfmx.o`  (not one single object file per executable).   This second approach means the duplication of any code between the application will not be handled by the linker so you could end up with much bigger final output (assuming there is some code sharing).

**Comment by vadimkantorov** [comment]
> When linking `.a` archive only the objects that are actually needed should be pulled in. Does the same build work on your host platform (linux, mac, windows etc)? This linking behaviour of emscripten is the same as on those platforms with respect the `.a` archive.

Good check. Will try and report



> If both copies of `jpegimage.o` are being pulled in then there must be some symbol from the second copy that is not present in the first copy (otherwise the second copy would no be linked in).

Normally two executables are built separately and are using their own modified copies of jpegimage (I checked, code is indeed slightly different).



> (caveat: Unless you are building with `-s MAIN_MODULE`.. are you doing that?)

Definitely not doing that :) Didn't know of MAIN_MODULE existence until now :)



> Having two different versions of `jpegimage` linked into the same program seems tricky at best and not a great idea. The best approach is probably going to be trying to remove the duplicate versions or at least ensure they are the same version.

I hope to overcome it somehow, unfortunately TexLive decided to use two different versions of jpegimage in-tree :(

> The second approach would be use `wasm-ld -r` (or `emcc -r`) to prelink each executable to a new, single object file and figure out a way to only export the main entry points from each one. Then your final link would be `emcc main.o busymain_xetex.o busymain_dvipdfmx.o`

Yep, that's what I tried. In the resulting busymain_xetex.o the jpeg symbols are still present :( with "*UND*". I was expecting they would be completely gone:
``` 
00000000       F *UND* check_for_jpeg
00000000       F *UND* check_for_bmp
00000000       F *UND* check_for_png
00297fb8 g     F CODE .hidden png_set_check_for_invalid_index
```
```
XETEX_OBJECTS="xetex-xetex-pool.o xetexdir/xetex-xetexextra.o synctexdir/xetex-synctex.o xetex-xetexini.o xetex-xetex0.o xetexdir/libxetex_a-XeTeXFontInst.o xetexdir/libxetex_a-XeTeXFontMgr.o xetexdir/libxetex_a-XeTeXLayoutInterface.o xetexdir/libxetex_a-XeTeXOTMath.o xetexdir/libxetex_a-XeTeX_ext.o xetexdir/libxetex_a-XeTeX_pic.o xetexdir/libxetex_a-trans.o xetexdir/libxetex_a-hz.o xetexdir/libxetex_a-pdfimage.o   xetexdir/libxetex_a-XeTeXFontMgr_FC.o"
XETEX_DEPS="$TEXLIVE_BUILD_DIR/libs/harfbuzz/libharfbuzz.a $TEXLIVE_BUILD_DIR/libs/graphite2/libgraphite2.a $TEXLIVE_BUILD_DIR/libs/teckit/libTECkit.a $TEXLIVE_BUILD_DIR/libs/libpng/libpng.a $TEXLIVE_BUILD_DIR/libs/freetype2/libfreetype.a $TEXLIVE_BUILD_DIR/libs/pplib/libpplib.a $TEXLIVE_BUILD_DIR/libs/zlib/libz.a libmd5.a lib/lib.a $TEXLIVE_BUILD_DIR/texk/kpathsea/.libs/libkpathsea.a $FONTCONFIG_BUILD_DIR/src/.libs/libfontconfig.a $EXPAT_BUILD_DIR/libexpat.a $TEXLIVE_BUILD_DIR/libs/icu/icu-build/lib/libicuuc.a $TEXLIVE_BUILD_DIR/libs/icu/icu-build/lib/libicudata.a"

pushd $TEXLIVE_BUILD_DIR/texk/web2c
/mnt/c/Users/user/emsdk/upstream/bin/wasm-ld  -relocatable $XETEX_OBJECTS $XETEX_DEPS -o test.o
```


I'm using `-Dmain='__attribute__((visibility("default"))) busymain_xetex'` to ensure that main gets exported

**Comment by sbc100** [comment]
It is curious that `check_for_jpeg` and others are not defined in the output of `wasm-ld  -relocatable`.  Can you trying with `--whole-archive` added before the `$XETEX_DEPS`? 

I thought that `-relocatable` would pull in what it needs from the archives.  Confirming that behaviour now.

**Comment by vadimkantorov** [comment]
It then fails with:
```
wasm-ld: error: duplicate symbol: main
>>> defined in /mnt/c/Users/user/xetex2020.js/texlive-source-9ed922e7d25e41b066f9e6c973581a4e61ac0328/texlive-build-wasm/libs/pplib/libpplib.a(pptest1.o)
>>> defined in /mnt/c/Users/user/xetex2020.js/texlive-source-9ed922e7d25e41b066f9e6c973581a4e61ac0328/texlive-build-wasm/libs/pplib/libpplib.a(pptest2.o)

wasm-ld: error: duplicate symbol: main
>>> defined in /mnt/c/Users/user/xetex2020.js/texlive-source-9ed922e7d25e41b066f9e6c973581a4e61ac0328/texlive-build-wasm/libs/pplib/libpplib.a(pptest2.o)
>>> defined in /mnt/c/Users/user/xetex2020.js/texlive-source-9ed922e7d25e41b066f9e6c973581a4e61ac0328/texlive-build-wasm/libs/pplib/libpplib.a(pptest3.o)

wasm-ld: error: duplicate symbol: main
>>> defined in /mnt/c/Users/user/xetex2020.js/texlive-source-9ed922e7d25e41b066f9e6c973581a4e61ac0328/texlive-build-wasm/libs/pplib/libpplib.a(pptest3.o)
>>> defined in lib/lib.a(main.o)

wasm-ld: error: duplicate symbol: argc
>>> defined in xetexdir/xetex-xetexextra.o
>>> defined in lib/lib.a(main.o)

wasm-ld: error: duplicate symbol: argv
>>> defined in xetexdir/xetex-xetexextra.o
>>> defined in lib/lib.a(main.o)
```

For first errors, I guess it somehow managed to put test binaries into the libpplib.a. I also wonder if argc/argv got defined as global symbols?

This is also strange because it only touches xetex dependencies, dvipdfmx is not linked in here.

**Comment by vadimkantorov** [comment]
When I remove libpplib.a and xetex-xetexextra.o from linking, `--whole-archive` succeeds, but still prints this:
```
00000000       F *UND* check_for_jpeg
00000000       F *UND* check_for_bmp
00000000       F *UND* check_for_png
002be8ac g     F CODE .hidden png_set_check_for_invalid_index
```

**Comment by vadimkantorov** [comment]
I fixed some bugs, some object files were missing from the linking command. However, even with `--whole-archive` command:
```
00155dd8 g     F CODE .hidden check_for_jpeg
00154a21 g     F CODE .hidden check_for_bmp
001544e0 g     F CODE .hidden check_for_png
```
symbols are hidden but still present. And hidden symbols still lead to duplicate symbols error if I understand it correctly

**Comment by vadimkantorov** [comment]
It seems that `wasm-ld` may support more custom export options: https://github.com/WebAssembly/tool-conventions/issues/64

**Comment by vadimkantorov** [comment]
I found sth like a surrogate for `--redefine-sym`: https://github.com/skalogryz/wasmbin. I'll let you know if it works

**Comment by sbc100** [comment]
> I fixed some bugs, some object files were missing from the linking command. However, even with `--whole-archive` command:
> 
> ```
> 00155dd8 g     F CODE .hidden check_for_jpeg
> 00154a21 g     F CODE .hidden check_for_bmp
> 001544e0 g     F CODE .hidden check_for_png
> ```
> 
> symbols are hidden but still present. And hidden symbols still lead to duplicate symbols error if I understand it correctly

Yes, hidden is the visibility which applies to the runtime dynamic linker.   if you want to avoid duplicate symbol errors you would need them them to have local binding.   I don't think there is any way to alter the binding of a symbol in wasm-ld so you would need some kind of external tool.

**Comment by vadimkantorov** [comment]
I managed to get around by manual `-Dold_symbol=new_symbol` renaming since I am in full control of compilation.

@sbc100 I suggest to reuse this issue to formulate a feature requst on improving wasm-ld or introducing some extra emscripten-specific function. Would you have an advice on wording?

**Comment by vadimkantorov** [comment]
Should this feature request go to LLVM? or leaving it here is fine?

**Comment by sbc100** [comment]
I don't think there is anything emscripten-specific about that problem you are trying to solve.

Let me see if I can formulate the problem you are trying to solve to see if I understand correctly the problem:

"I'd like to be able to link a single application that is made up of sources from multiple different applications that were not originally intended to be linked as one.  In order to avoid symbol collisions between these sub-applications I would like to internalize the symbols of each one before linking them together". 

Does that sound about right?

To me that sounds like a job for objcopy (or llvm-objcopy) which is the tool to modifying object files before feeding them to the linker.     In fact I think there might already be an option to do what you want: `--localize-hidden`.    See https://stackoverflow.com/questions/393980/restricting-symbols-in-a-linux-static-library.

The linking process would then be:

1. Link each sub-application to an object file `wasm-ld -r <obj_files..> -o app1.o`
2. Convert all hidden symbol to local: `llvm-objcopy --localize-hidden app1.o -o app1_localized.o
3. Link all the localized apps together (this time without -r). `wasm-ld app1_localized.o ... -o final.wasm`

If that sounds like accurate problem description and a reasonable solution the main blocker is that `llvm-objcopy` currently has very limited support for wasm object format.    So I would start by opening an llvm bug requesting `--localize-hidden` be implemented in `llvm-objcopy` for the wasm format.   If you are willing it might be reasonable bug to try to work on yourself.

**Comment by vadimkantorov** [comment]
Indeed, if llvm-objcopy supports `--localize-hidden`, it may be solved. I'll check.
Another way would be to support `ld`'s `--redefine-sym` in `wasm-ld` as well.

**Comment by sbc100** [comment]
llvm-objcopy almost certainly does not support `--localize-hidden` for the wasm object format since support for that object format is very limited today.  The flag itself is supported, but work would most likely be needed to make it work with wasm-ld.

I don't think redefine symbol is a great solution here because it requires to to go through and make a list of all possible colliding symbols and pass each one on the command line.  Were you also thinking of using that with `-r` when pre-linking each sub-application?   I means changes to the underlying source code can break your build.

**Comment by sbc100** [comment]
Also, I don't see a `--redefine-sym` in GNU ld or in lld.. where is that reference coming from?  Are you talking about objcopy maybe?

**Comment by vadimkantorov** [comment]
> . Were you also thinking of using that with `-r` when pre-linking each sub-application?

I'll check what `-r` is!

> I don't think redefine symbol is a great solution

Agreed, but it's a better solution than crafting defines (defines are even more brittle)

> Also, I don't see a `--redefine-sym` in GNU ld or in lld.

You're right, it's also coming from `objcopy`

**Comment by vadimkantorov** [comment]
> I don't think there is anything emscripten-specific about that problem you are trying to solve.
> 
> Let me see if I can formulate the problem you are trying to solve to see if I understand correctly the problem:
> 
> "I'd like to be able to link a single application that is made up of sources from multiple different applications that were not originally intended to be linked as one. In order to avoid symbol collisions between these sub-applications I would like to internalize the symbols of each one before linking them together".
> 
> Does that sound about right?
> 
> To me that sounds like a job for objcopy (or llvm-objcopy) which is the tool to modifying object files before feeding them to the linker. In fact I think there might already be an option to do what you want: `--localize-hidden`. See https://stackoverflow.com/questions/393980/restricting-symbols-in-a-linux-static-library.
> 
> The linking process would then be:
> 
> 1. Link each sub-application to an object file `wasm-ld -r <obj_files..> -o app1.o`
> 2. Convert all hidden symbol to local: `llvm-objcopy --localize-hidden app1.o -o app1_localized.o
> 3. Link all the localized apps together (this time without -r). `wasm-ld app1_localized.o ... -o final.wasm`
> 
> If that sounds like accurate problem description and a reasonable solution the main blocker is that `llvm-objcopy` currently has very limited support for wasm object format. So I would start by opening an llvm bug requesting `--localize-hidden` be implemented in `llvm-objcopy` for the wasm format. If you are willing it might be reasonable bug to try to work on yourself.

I finally created an account at LLVM tracker. What "product" should I file the issue in? lld? clang? projects?

```
Bugzilla Admin:	Things we want to do to enhance our Bugzilla experience.
Build scripts:	CMake, Makefiles and autoconf
clang:	The 'clang' C family front-end
clang-tools-extra:	Extra tools based on Clang
compiler-rt:	The compiler runtime library.
Documentation:	The documentation for the LLVM system.
dragonegg:	GCC plugin that replaces the GCC optimizer and code generator with LLVM.
DSA:	The Data Structure Analysis project.
flang:	The flang Fortran front end.
klee:	klee symbolic virtual machine: http://klee.llvm.org/
libc++:	The libc++ C++ Standard Library
libc++abi:	libc++abi is a new implementation of low level support for a standard C++ library.
libclc:	An implementation of the library requirements of the OpenCL C programming language
libraries:	All of the LLVM libraries (most bugs end up here)
lld:	The LLVM Linker project, http://lld.llvm.org/
lldb:	The LLVM Debugger project, http://lldb.llvm.org/
llgo:	LLVM-based compiler for Go
LNT:	The LLVM performance tracking software.
MLIR:	https://mlir.llvm.org/
new-bugs:	File new bugs here
OpenMP:	OpenMP Runtime Library & Compiler support
Packaging:	Packaging and distribution related issues
parallel STL:	parallel STL
Phabricator:	LLVM Phabricator
Polly:	Polly is a high-level loop and data-locality optimizer and optimization infrastructure for LLVM.
PoolAlloc:	The Automatic Pool Allocation project.
Projects:	Various projects that live in the LLVM sourcebase, or are otherwise tied into LLVM
Runtime Libraries:	These are all of the runtime libraries that live in llvm/runtime
SAFECode:	The SAFECode project (http://safecode.cs.illinois.edu/).
Test Suite:	Tools for testing LLVM
tools:	LLVM toolchain
Website:	Various websites that are maintained by our group
XRay:	XRay function call tracing with compiler instrumentation and dynamic runtime control
```

**Comment by vadimkantorov** [comment]
@sbc10 Created https://bugs.llvm.org/show_bug.cgi?id=51279 in new-bugs

It's my first LLVM bug, so please advise me if I can tag you there or otherwise add some useful metadata to this report :)

**Comment by vadimkantorov** [comment]
@sbc100 By the way, https://github.com/skalogryz/wasmbin#wasmtool complains that `wat2wasm` produces un-linkable WASM binaries

**Comment by sbc100** [comment]
Perhaps raise a bug with https://github.com/skalogryz/wasmbin#wasmtool ?  Or with wabt if you think the problem lies there?

**Comment by vadimkantorov** [comment]
If I understand well the README in this repo, they complain about `wabt`, yeah

**Comment by sbc100** [comment]
I'm not sure how that is relevant here in the emscripten repo?    It sounds like maybe somebody (presumably the wasmtool author) should open a wabt issue?

**Comment by vadimkantorov** [comment]
I am just wondering if that complain makese sense to you (assuming you're aware of state of the art of those tools). Yep, I'll open an issue there

**Comment by sbc100** [comment]
Hmm... I can't see any mention of wabt in the https://github.com/skalogryz/wasmbin#wasmtool README so its hard to say, but there could well me issues with `wat2wasm -r`.   I think that feature/flag doesn't get much use/attention.   I'd certainly be interested if they can supply a repro case.

**Comment by vadimkantorov** [comment]
The mention I meant is
```
wat2wasm util is capable of generating binaries with relocation information. However the symbol flags are setup in the manner that final binaries cannot be linked by wasm-ld utility. (i.e. non setting weak symbols).
```

**Comment by sbc100** [comment]
Oh I see.  They are basically saying why you might want to use their tool.  wat format doesn't support such things as marking symbols as weak so if you need weak symbls you can use their tool to add them.  I think they are describing a use case rather than a bug in wat2wasm.

**Comment by vadimkantorov** [comment]
I guess it's also a case that you can't reliably use wasm2wat / wat2wasm for renaming colliding symbols...

**Comment by sbc100** [comment]
You can't using `wasm2wat` + `wat2wasm` to round trip object files, no.   Some of the information in the object file will be lost when you run `wasm2wat` since the wat format can't represent the linking information.

**Comment by kon72** [comment]
How about using `obj2yaml` + `yaml2obj` to rename symbols?

**Comment by sbc100** [comment]
I don't think we currently ship obj2yaml or yaml2obj in emsdk.  My understanding is that those tools are mostly for llvm testing, but maybe I'm wrong about that?

**Comment by kon72** [comment]
I agree with you that shipping obj2yaml and yaml2obj is not the right way, but I'm just wondering if there is a reliable way to do symbol renaming when there is no official tooling for that.

**Comment by sbc100** [comment]
IIUC the state of this issue is:

1.  There is no official way to do this today (but there are tools such as wasm2wat + wat2wasm and  obj2yaml and yaml2obj can could probably do it in a hacky way).
2. We hope to add support for `--redefine-sym` and `--localize-hidden` to the Wasm backend of llvm-objcopy in the future.

**Comment by vadimkantorov** [comment]
@sbc100 @kripken My friend @dfyz joined me in the texlive ❤️ wasm (aka busytex) endeavor and implemented wasm symbol renaming into llvm!

PR incoming, any llvm/wasm guidance to discuss/facilitate a merge will be very welcome :) 

https://github.com/llvm/llvm-project/issues/50623#issuecomment-1980057649
